adding CI workflow for creating mpv binary and debian `.deb` package

ilkermeliksitki opened the pull request:
Currently, the workflow is only running on `ubuntu-24.04`. 

- a working mpv binary file is gotten
- a debian package is created

During the creation of `.deb` package, `libgnutls-dev` couldn't be located and instead the other alternative `libgnutls28-dev` is used. I am not sure, but maybe we can delete it from the dependencies inside the `debian/control` file.

ilkermeliksitki committed:
add workflow to build mpv binary

Changes to be committed:
	new file:   build_mpv_binary.yml

ilkermeliksitki committed:
minor fixes

1- change the name of the workflow
2- change the branch name from main to master
3- add workflow_dispatch

Changes to be committed:
	modified:   build_mpv_binary.yml

ilkermeliksitki committed:
fix syntax error by adding backslash

Changes to be committed:
	modified:   build_mpv_binary.yml

ilkermeliksitki committed:
add the remaining dependencies

Changes to be committed:
	modified:   build_mpv_binary.yml

ilkermeliksitki committed:
delete trailing space

ilkermeliksitki committed:
add sudo to `apt clean` command

Changes to be committed:
	modified:   build_mpv_binary.yml

ilkermeliksitki committed:
add install step and change the verifying command

Changes to be committed:
	modified:   build_mpv_binary.yml

ilkermeliksitki committed:
first attempt to create `.deb` package for mpv

Changes to be committed:
	modified:   build_mpv_binary.yml

ilkermeliksitki committed:
add checking mechanism for the existence of `.deb` file

Changes to be committed:
	modified:   build_mpv_binary.yml

ilkermeliksitki committed:
rename the workflow and make it more descriptive

Changes to be committed:
	renamed:    build_mpv_binary.yml -> build_mpv_binary_and_deb_package.yml

ilkermeliksitki committed:
use `apt-get` instead of `apt` in the script

Changes to be committed:
	modified:   build_mpv_binary_and_deb_package.yml

sfan5 commented on .github/workflows/build_mpv_binary_and_deb_package.yml:
```diff
@@ -0,0 +1,139 @@
+            libxv-dev \
+            libjpeg-dev \
+            libssl-dev \
+            libx264-dev \
+            libmp3lame-dev \
+            libfdk-aac-dev \
```
encoders are not needed either

ilkermeliksitki replied:
the encoders `libmp3lame-dev, libx264-dev` are necessary for creating `.deb` package, that's why I added. If I don't I get the following error. But I will create a separate workflow for building the debian package as you have suggested.

```
dpkg-buildpackage: info: host architecture amd64
dpkg-source: info: using options from mpv-build/debian/source/options: --compression=xz --compression-level=best
dpkg-checkbuilddeps: error: Unmet build dependencies: libmp3lame-dev libx264-dev
dpkg-buildpackage: warning: build dependencies/conflicts unsatisfied; aborting
dpkg-buildpackage: warning: (Use -d flag to override.)
```

sfan5 commented on .github/workflows/build_mpv_binary_and_deb_package.yml:
```diff
@@ -0,0 +1,139 @@
+            libmp3lame-dev \
+            libfdk-aac-dev \
+            python3-pip \
+            python3.12-venv \
```
do we use venv or pip anywhere?

ilkermeliksitki replied:
No actually. I used these lines to download `youtube-dl` by using python, because the package was not available in the ubuntu repositories but it is only a run time dependency. I guess we can delete these packages safely. Thank you.

sfan5 commented on .github/workflows/build_mpv_binary_and_deb_package.yml:
```diff
@@ -0,0 +1,139 @@
+            libharfbuzz-dev \
+            nasm \
+            liblua5.1-0-dev \
+            liblua5.2-dev \
```
one lua version is enough

sfan5 commented on .github/workflows/build_mpv_binary_and_deb_package.yml:
```diff
@@ -0,0 +1,139 @@
+            libxinerama-dev \
+            libxv-dev \
+            libwayland-dev \
+            nodejs \
```
nodejs is definitely not needed

ilkermeliksitki replied:
after running `./rebuild -j4` command, I get the following result. 

```
mpv 0.39.0-UNKNOWN

    d3d11     : NO
    javascript: YES
    libmpv    : NO
    lua       : YES
    opengl    : YES
    vulkan    : YES
    wayland   : YES
    x11       : YES
   ```

To turn `javascript` option to `YES`, I uploaded `nodejs`. But I can certainly delete if it is needed at all.

avih replied:
> To turn `javascript` option to `YES`, I uploaded `nodejs`

mpv's js doesn't use nodejs. It uses mujs, which is a tiny interpreter.

ilkermeliksitki replied:
@avih thank you!

sfan5 commented on .github/workflows/build_mpv_binary_and_deb_package.yml:
```diff
@@ -0,0 +1,139 @@
+            libpipewire-0.3-dev \
+            libpulse-dev \
+            libsndio-dev \
+            libcaca-dev \
```
mpv support libcaca but it's really not important

sfan5 commented on .github/workflows/build_mpv_binary_and_deb_package.yml:
```diff
@@ -0,0 +1,139 @@
+
+    - name: Install mpv binary
+      run: |
+        sudo ./install
```
no need to install it just to verify that it works

sfan5 commented on .github/workflows/build_mpv_binary_and_deb_package.yml:
```diff
@@ -0,0 +1,139 @@
+        mpv --version
+
+    - name: Create debian package
+      run: |
```
I'd move this to a different CI job

ilkermeliksitki replied:
Alright. I will create a separate CI job. Thank you.

sfan5 commented on .github/workflows/build_mpv_binary_and_deb_package.yml:
```diff
@@ -0,0 +1,139 @@
+        # enable source repositories
+        sudo sed -i 's/^Types: deb/Types: deb deb-src/g' /etc/apt/sources.list.d/ubuntu.sources
+        sudo apt-get update
+        # install dependencies mentioned in debian/control
```
is there no way to say "take this control file and install everything mentioned there"? a bit unfortunate to duplicate the list here

wsldankers replied:
Yes. You can do something like:
```sh
sudo apt-get install -y build-essential devscripts equivs
mk-build-deps -s sudo -i
```
You can also remove the `apt-get build-dep` if you do that.

ilkermeliksitki replied:
@wsldankers I am currently testing the following command on my local system.

```
equivs-build --full debian/control
```

Which approach is better you think?

wsldankers replied:
That does something else, I think. Do you have an objection to the original suggestion?

ilkermeliksitki replied:
No, I don't have an objection and probably you are right.  After running that command I get `mpv_1.0_amd64.deb` file and I thought it was the right one.

sfan5 commented on .github/workflows/build_mpv_binary_and_deb_package.yml:
```diff
@@ -0,0 +1,139 @@
+            && sudo apt-get clean
+
+        # build dependencies
+        sudo apt-get build-dep mpv
```
this *installs* build dependencies but I am pretty sure this references whatever dependencies the `mpv` repo package has and not what's inside `debian/control`

sfan5 commented on .github/workflows/build_mpv_binary_and_deb_package.yml:
```diff
@@ -0,0 +1,139 @@
+        sudo dpkg-buildpackage -us -uc
+
+        # check if package is created
+        if [ -f ../mpv*.deb ]; then
+            echo "mpv debian package creation is successful"
+        else
+            echo "mpv debian package creation failed!"
+            exit 1
+        fi
```
```suggestion
        ls -lh ../mpv*.deb || exit 1
```

sfan5 commented:
Can I see the workflow in action somewhere?

ilkermeliksitki committed:
Update .github/workflows/build_mpv_binary_and_deb_package.yml

add `actions/checkout@v4`

Co-authored-by: sfan5 <sfan5@live.de>

ilkermeliksitki committed:
delete `clang`

Changes to be committed:
	modified:   build_mpv_binary_and_deb_package.yml

ilkermeliksitki committed:
simplify the step of checking debian package

Co-authored-by: sfan5 <sfan5@live.de>

ilkermeliksitki commented:
> Can I see the workflow in action somewhere?

@sfan5  I guess you should be able to see the logs [here](https://github.com/ilkermeliksitki/mpv-build/actions).

ilkermeliksitki committed:
delete `python3-pip` and `python3.12-venv`

Changes to be committed:
	modified:   build_mpv_binary_and_deb_package.yml

ilkermeliksitki committed:
delete one of the `lua` versions

Changes to be committed:
	modified:   build_mpv_binary_and_deb_package.yml

ilkermeliksitki committed:
delete unnecessary encoders

Changes to be committed:
	modified:   build_mpv_binary_and_deb_package.yml

ilkermeliksitki committed:
delete `nodejs`

Changes to be committed:
	modified:   build_mpv_binary_and_deb_package.yml

ilkermeliksitki committed:
delete `libcaca`

Changes to be committed:
	modified:   build_mpv_binary_and_deb_package.yml

ilkermeliksitki committed:
add `libmp3lame-dev` and `libx264-dev`

Changes to be committed:
	modified:   build_mpv_binary_and_deb_package.yml

ilkermeliksitki committed:
replace debian/compat with debhelper-compat

While building the package, by using equivs-build and debian/control
file, I got the following error:

```
root@a625b7a1de7c:/mpv-build# equivs-build --full debian/control
dpkg-buildpackage: info: source package mpv
dpkg-buildpackage: info: source version 1.0
dpkg-buildpackage: info: source distribution unstable
dpkg-buildpackage: info: source changed by Kevin Mitchell <kevmitch@gmail.com>
dpkg-buildpackage: info: host architecture amd64
 dpkg-source --before-build .
 debian/rules clean
dh clean
dh: error: Please specify the compatibility level in debian/compat or via Build-Depends: debhelper-compat (= X)
make: *** [debian/rules:3: clean] Error 255
dpkg-buildpackage: error: debian/rules clean subprocess returned exit status 2
Error in the build process: exit status 2
```
Based on the error message, I replaced debian/compat with
debhelper-compate (= 12) in debian/control file. The existing
compat file didn't help to build the package.

[Reference](https://github.com/FooBarWidget/debian-packaging-for-the-modern-developer/issues/8#issue-665598885)

Changes to be committed:
      deleted:    compat
      modified:   control

wsldankers commented:
In general, is it useful to build the package twice, once separately and once as a package? Building the package already exercises the build scripts. We could drop the first two steps and if we want to run mpv --version we could just install the resulting .deb file.

avih commented:
Can you please clarify the goal of this PR?

Is it to test that a normal build works, and also the debian packging works?

Or maybe only one of those?

Is it supposed to produce an artifact which debian users can install?

Something else?

ilkermeliksitki commented:
@avih My aim is to create a workflow that does:

1. one that just runs the scripts: we should get a functioning mpv binary
2. one that invokes the debian packaging stuff: we should get a deb package

Here is the original [reference](https://github.com/mpv-player/mpv/issues/13608#issuecomment-2525916922) given by @sfan5.

ilkermeliksitki committed:
Revert "replace debian/compat with debhelper-compat"

This reverts commit a664c8d604466ce62f0ab93ac058975c9482e172.

Changes to be committed:
	new file:   debian/compat
	modified:   debian/control

ilkermeliksitki committed:
suggested changes applied

1. installation step is deleted during binary building
2. building `.deb` package is moved to `build_debian_package.yml`
3. `build_mpv_binary_and_deb_package.yml` is renamed to `build_debian_package.yml`
Changes to be committed:
      renamed:    build_mpv_binary_and_deb_package.yml -> build_debian_package.yml
      new file:   build_mpv_binary.yml

ilkermeliksitki committed:
check the existence of `.deb` file

Changes to be committed:
	modified:   build_debian_package.yml

ilkermeliksitki committed:
delete enable source repository part

Changes to be committed:
	modified:   build_debian_package.yml

ilkermeliksitki committed:
build-essential package is already installed

Changes to be committed:
	modified:   build_debian_package.yml

sfan5 reviewed (commented):
Looks good otherwise, where can I see them in action?

sfan5 commented on .github/workflows/build_debian_package.yml:
```diff
@@ -0,0 +1,90 @@
+        sudo apt-get install -y devscripts equivs
+
+        # create debian package
+        mk-build-deps -s sudo -i
```
looking at https://github.com/ilkermeliksitki/mpv-build/actions/runs/12402609257/job/34624456632 this seems to create some sort of dependency package, but it doesn't actually build mpv

ilkermeliksitki committed:
debian package build attempt 1
Changes to be committed:
	modified:   build_debian_package.yml

ilkermeliksitki committed:
add the run key for `verify package creation` step

Changes to be committed:
	modified:   build_debian_package.yml

ilkermeliksitki committed:
add libmpv-dev library

during the build procedure, the following error was encountered:

```
...
 [303/307] Compiling C object libmpv.so.2.5.0.p/video_out_vo_vdpau.c.o
[304/307] Linking target libmpv.so.2.5.0
FAILED: libmpv.so.2.5.0
cc  -o libmpv.so.2.5.0 libmpv.so.2.5.0.p/audio_aframe.c.o libmpv.so.2.5.0.p/audio_chmap.c.o libmpv.so.2.5.0.p/audio_chmap_avchannel.c.o libmpv.so.2.5.0.p/audio_chmap_sel.c.o libmpv.so.2.5.0.p/audio_decode_ad_lavc.c.o libmpv.so.2.5.0.p/audio_decode_ad_spdif.c.o libmpv.so.2.5.0.p/audio_filter_af_drop.c.o libmpv.so.2.5.0.p/audio_filter_af_format.c.o libmpv.so.2.5.0.p/audio_filter_af_lavcac3enc.c.o libmpv.so.2.5.0.p/audio_filter_af_scaletempo.c.o libmpv.so.2.5.0.p/audio_filter_af_scaletempo2.c.o libmpv.so.2.5.0.p/audio_filter_af_scaletempo2_internals.c.o libmpv.so.2.5.0.p/audio_fmt-conversion.c.o libmpv.so.2.5.0.p/audio_format.c.o libmpv.so.2.5.0.p/audio_out_ao.c.o libmpv.so.2.5.0.p/audio_out_ao_lavc.c.o libmpv.so.2.5.0.p/audio_out_ao_null.c.o libmpv.so.2.5.0.p/audio_out_ao_pcm.c.o libmpv.so.2.5.0.p/audio_out_buffer.c.o libmpv.so.2.5.0.p/common_av_common.c.o libmpv.so.2.5.0.p/common_av_log.c.o
...```

Changes to be committed:
	modified:   build_mpv_binary.yml

ilkermeliksitki committed:
Revert "add libmpv-dev library"

This reverts commit b0a19d74cbddfcb7b69e1e13bb45d535a3b9e6de.

ilkermeliksitki committed:
Fix trial of linkage error of libass during "rebuild" step

Changes to be committed:
	modified:   scripts/libass-build

ilkermeliksitki committed:
Revert "Fix trial of linkage error of libass during "rebuild" step"

This reverts commit cce2eb6f9e2f1902129b74c9eafdd0ecde403b24.
The trial didn't work.

Changes to be committed:
	modified:   scripts/libass-build

ilkermeliksitki committed:
Merge branch 'mpv-player:master' into master

ilkermeliksitki committed:
add pwd for debugging purposes

Changes to be committed:
	modified:   .github/workflows/build_debian_package.yml

ilkermeliksitki committed:
change update.sh to update

Changes to be committed:
	modified:   .github/workflows/build_debian_package.yml

ilkermeliksitki committed:
add devscript to build debian package

devscripst include the following command to be used in the
workflow:
  - dpkg-buildpackage
  - mk-build-deps

Changes to be committed:
	modified:   .github/workflows/build_debian_package.yml

ilkermeliksitki committed:
add equivs package

The workflow gave the following error:
  mk-build-deps: You must have equivs installed to use this program.
  Error: Process completed with exit code 127.
So equivs package is added as dependency.

Changes to be committed:
	modified:   .github/workflows/build_debian_package.yml

ilkermeliksitki committed:
delete the debuggin pwd line

Changes to be committed:
	modified:   .github/workflows/build_debian_package.yml

ilkermeliksitki reviewed (commented):
All feedback are addressed. CI is working. Both building mpv binary, and debian package.

kasper93 commented on .github/workflows/build_debian_package.yml:
```diff
@@ -0,0 +1,90 @@
+
+    - name: Build Debian package
+      run: |
+        dpkg-buildpackage -uc -us -b -j4
```
```suggestion
        dpkg-buildpackage -uc -us -b -j`nproc`
``` 

don't assume number of threads.

kasper93 commented on .github/workflows/build_debian_package.yml:
```diff
@@ -0,0 +1,90 @@
+            libshaderc-dev \
+            devscripts \
+            equivs \
+            && sudo apt-get clean
```
I don't think it's worth taking time to clean, the whole VM/container will be destroyed after job anyway.

kasper93 commented on .github/workflows/build_debian_package.yml:
```diff
@@ -0,0 +1,90 @@
+    - name: Install dependencies
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y \
```
```suggestion
        sudo apt-get install --no-install-recommends -y \
```
By default it installs a lot of bloat that we don't need and likely it's better to manual control what packages we have.

kasper93 commented on .github/workflows/build_debian_package.yml:
```diff
@@ -0,0 +1,90 @@
+  pull_request:
+    branches:
+      - master
+  workflow_dispatch:
```
Have you considered adding periodic build?
For example we could build this every day at midnight or every week.

```
  schedule:
    - cron: "0 0 * * *"
```

ilkermeliksitki replied:
only for this `build_debian_package.yml` or should I do it for `build_mpv_binary.yml` as well?

ilkermeliksitki committed:
the number of thread is not assumed

Co-authored-by: Kacper Michajłow <kasper93@gmail.com>

ilkermeliksitki committed:
Update .github/workflows/build_debian_package.yml

Co-authored-by: Kacper Michajłow <kasper93@gmail.com>

ilkermeliksitki committed:
run the action every day

Changes to be committed:
	modified:   build_debian_package.yml

ilkermeliksitki committed:
delete apt-get clean

Changes to be committed:
	modified:   build_debian_package.yml

ilkermeliksitki committed:
delete apt clear for mpv build

Changes to be committed:
	modified:   build_mpv_binary.yml

ilkermeliksitki committed:
use all available cores in rebuild instead of fixed j4

Changes to be committed:
	modified:   build_mpv_binary.yml

sfan5 commented on .github/workflows/build_debian_package.yml:
```diff
@@ -0,0 +1,85 @@
+            libva-dev \
+            libshaderc-dev \
+            devscripts \
+            equivs
```
can we automatically take the dependencies from `debian/control` instead of listing them here?
isn't this what `mk-build-deps` is for?

ilkermeliksitki replied:
yes indeed @sfan5. Thank you!. I used `mk-builds-deps` instead of hardcoded ones.

ilkermeliksitki committed:
Merge branch 'mpv-player:master' into master

ilkermeliksitki committed:
refactor: ci workflow for debian package build and fix build

1- Removes the obsolete `-Dsdl2=enabled` flag from debian/rules
   to fix the Meson configuration error.
2- Refactors harcoded dependency list for the debian package

Changes to be committed:
      modified:   .github/workflows/build_debian_package.yml
      modified:   debian/rules

ilkermeliksitki commented on .github/workflows/build_debian_package.yml:
```diff
@@ -0,0 +1,85 @@
+
+    - name: Verify package creation
+      run: |
+        ls ../mpv_*.deb || exit 1
```
may I ask why is this necessary?

sfan5 replied:
it's nice to see how big the resulting deb file is
you can add `-h` too

ilkermeliksitki replied:
Thank you. It is added.

sfan5 commented:
Where can I see a test run of the CI working?

ilkermeliksitki committed:
improve logging by showing the resulting deb files size

- also, `-h` flag is added for human-readable sizes

Changes to be committed:
      modified:   .github/workflows/build_debian_package.yml

ilkermeliksitki commented:
> Where can I see a test run of the CI working?

@sfan5, you can see from here, [link](https://github.com/ilkermeliksitki/mpv-build/actions)

Also, currently building mpv binary action is triggered by pushing and pulling. Should I change it to daily basis as well like building debian package action?
